stacks-signer/src/runloop.rs

```diff
@@ -13,1607 +13,320 @@
-                match self.coordinator.start_dkg_round() {
-                    Ok(msg) => {
-                        let ack = self.stackerdb.send_message_with_retry(msg.into());
-                        debug!("Signer #{}: ACK: {:?}", self.signer_id, ack);
-                        self.state = State::Dkg;
-                        true
+    /// Refresh the signer configuration by retrieving the necessary information from the stacks node
+    /// Note: this will trigger DKG if required
+    fn refresh_signers_with_retry(&mut self, current_reward_cycle: u64) -> Result<(), ClientError> {
+        let next_reward_cycle = current_reward_cycle.saturating_add(1);
+        retry_with_exponential_backoff(|| {
+            if let Err(e) = self.refresh_signer_config(current_reward_cycle) {
+                match e {
+                    ClientError::NotRegistered => {
+                        debug!("Signer is NOT registered for the current reward cycle {current_reward_cycle}.");
                     }
-                    Err(e) => {
-                        error!("Failed to start DKG: {:?}", e);
-                        warn!("Resetting coordinator's internal state.");
-                        self.coordinator.reset();
-                        false
+                    ClientError::RewardSetNotYetCalculated(_) => {
+                        debug!("Current reward cycle {current_reward_cycle} reward set is not yet calculated. Let's retry...");
+                        return Err(backoff::Error::transient(e));
                     }
+                    _ => return Err(backoff::Error::transient(e)),
                 }
             }
-            RunLoopCommand::Sign {
-                block,
-                is_taproot,
-                merkle_root,
-            } => {
-                let signer_signature_hash = block.header.signer_signature_hash();
-                let block_info = self
-                    .blocks
-                    .entry(signer_signature_hash)
-                    .or_insert_with(|| BlockInfo::new(block.clone()));
-                if block_info.signed_over {
-                    debug!("Signer #{}: Received a sign command for a block we are already signing over. Ignore it.", self.signer_id);
-                    return false;
-                }
-                info!("Signer #{}: Signing block: {:?}", self.signer_id, block);
-                match self.coordinator.start_signing_round(
-                    &block.serialize_to_vec(),
-                    *is_taproot,
-                    *merkle_root,
-                ) {
-                    Ok(msg) => {
-                        let ack = self.stackerdb.send_message_with_retry(msg.into());
-                        debug!("Signer #{}: ACK: {:?}", self.signer_id, ack);
-                        self.state = State::Sign;
-                        block_info.signed_over = true;
-                        true
+            if let Err(e) = self.refresh_signer_config(next_reward_cycle) {
```

Comment:
Seems you're calling `refresh_signer_config` twice?

Reply:
Yes one is for the current and one for the next reward cycle.